atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/locks/Lockable.java

```diff
@@ -0,0 +1,157 @@
+
+public final class Lockable<T> {
+    private static final SafeLogger log = SafeLoggerFactory.get(Lockable.class);
+    private final LockDescriptor lockDescriptor;
+    private final T inner;
+    private final TimelockService timelockService;
+    private final Refreshable<Duration> lockTimeout;
+
+    // Consider whether this should be a lock. It's simply an optimisation to avoid a check for timelock
+    // each time you want to test if it's locked by something locally, but we must make sure there are no false
+    // negatives. It's allowed that something local is working on this item and Timelock has released the lock due to
+    // expiry and this is still True (since that still lets us minimise conflicts by not working on it locally),
+    // but it should not be true if something local has stopped working on this item and closed the inner class
+    // basically, if this is locked then timelock may have the lock token active,
+    // but if this is unlocked timelock does not have the lock token active.
+    private final AtomicBoolean isLocked = new AtomicBoolean(false);
+
+    private Lockable(
+            T inner,
+            LockDescriptor lockDescriptor,
+            TimelockService timelockService,
+            Refreshable<Duration> lockTimeout) {
+        this.timelockService = timelockService;
+        this.inner = inner;
+        this.lockTimeout = lockTimeout;
+        this.lockDescriptor = lockDescriptor;
+    }
+
+    public static <T> Lockable<T> create(
+            T inner, LockDescriptor lockDescriptor, TimelockService timeLock, Refreshable<Duration> lockTimeout) {
+        return new Lockable<>(inner, lockDescriptor, timeLock, lockTimeout);
+    }
+
+    // For code owners:
+    // Do not use this for multiple locks!! If you're not careful, you can make it really easy to deadlock
+    // If you end up using this class and have a need for batch locks, implement a separate method that uses
+    // Timelocks batch locking methods and ensure there's a consistent ordering when taking out local locks
+    // to avoid circular waiting.
+    //    @MustBeClosed TODO: How do I get this working?? I'll track looking into this for another time.
+    @CheckReturnValue
+    public Optional<LockedItem<T>> tryLock(Consumer<Lockable<T>> disposeCallback) {
+        if (isLocked.compareAndSet(false, true)) {
+            // We do not want the timeout to be too low to avoid a race condition where we give up too soon
+            LockRequest request = LockRequest.of(
+                    ImmutableSet.of(lockDescriptor), lockTimeout.get().toMillis());
+
+            try {
+                return timelockService
+                        .lock(request)
+                        .getTokenOrEmpty()
+                        .map(lockToken ->
+                                new LockedItem<>(lockToken, this, inner, disposeCallback, timelockService, isLocked));
```

Comment:
An alternative to the parameter mess is to pass in a _single_ close method is something like 

```
Runnable unlock(Consumer<Lockable<T> disposeCallback) {
            try {
                timelockService.unlock(ImmutableSet.of(lockToken));
            } finally {
                isLocked.set(false); // We end up holding the local lock ever so slightly longer than necessary.
                disposeCallback.accept(wrappingLockable);
            }
}
...
new LockedItem<>(item, this::unlock)
```

which still captures `this` as a strong reference, which is sufficient for the LockableFactory stuff.